addons/website_sale/models/website.py

```diff
@@ -666,3 +675,22 @@ def _get_checkout_steps(self, current_step=None):
         if current_step:
             return next(step for step in steps if current_step in step[0])[1]
         return steps
+
+    def check_shop_access(self):
+        if self.is_public_user() and self.ecommerce_access == 'logged':
+            return request.redirect('/web/login')
+
+
+class Menu(models.Model):
+    _inherit = "website.menu"
+
+    def _compute_visible(self):
+        """ Hide '/shop' menus if ecommerse is allowed only to logged users and current user is
+        public."""
+        shop_menus = self.filtered(lambda menu: menu.url[:9] == "/shop")
+
+        for menu in shop_menus:
+            menu.is_visible = not (
+                    menu.website_id.ecommerce_access == 'logged' and request.env.user._is_public()
+            )
```

Comment:
```suggestion
            menu.is_visible =  menu.website_id.has_shop_access()
```

Reply:
This cannot be done using has_shop_access function.